Caching of the construction of constitutive laws in the model framework

keileg opened the pull request:
## Proposed changes
As discussed in #1345, the cost of constructing operators for constitutive laws is high, in particular for laws that are called multiple times. This PR introduces a simple caching-mechanism that stores the constructed operator.

Caching is invoked by adding a decorator `@pp.ad.cached_method` on target methods. Though in principle caching can be applied widely, this PR introduces it only in a few selected methods. Profiling indicates that this PR, together with an upcoming PR that targets parameter updates within Newton iterations, removes most if not all the non-standard bottlenecks (those that are not assembly linear solve, rediscretization, exporting) within Newton.

---------------------

UPDATE: Preserving type information for the decorated methods turned out to be non-trivial (for reference: Using `functools.wraps()` should solve the issue, but in practice it does not - I can speculate that the combination of many classes, mixins and decorators is simply too much for mypy). Having explored all reasonable suggestions from the internet and copilot, I had to introduce the cached methods in the protocol. This is far from a perfect solution, but I am afraid it will have to do.


## Types of changes

What types of changes does this PR introduce to PorePy?
_Put an `x` in the boxes that apply._

- [ ] Minor change (e.g., dependency bumps, broken links).
- [ ] Bugfix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] Testing (contribution related to testing of existing or new functionality).
- [ ] Documentation (contribution related to adding, improving, or fixing documentation).
- [x] Maintenance (e.g., improve logic and performance, remove obsolete code).
- [ ] Other:

## Checklist

_Put an `x` in the boxes that apply or explain briefly why the box is not relevant._

- [ ] The documentation is up-to-date.
- [ ] Static typing is included in the update.
- [ ] This PR does not duplicate existing functionality.
- [ ] The update is covered by the test suite (including tests added in the PR).
- [ ] If new skipped tests have been introduced in this PR, `pytest` was run with the `--run-skipped` flag.

IvarStefansson reviewed (changes requested):
What has been done looks good. But I don't see any testing of the caching. I think that would be wise.

keileg commented on src/porepy/models/constitutive_laws.py:
```diff
@@ -105,6 +107,7 @@ def elastic_displacement_jump(self, subdomains: list[pp.Grid]) -> pp.ad.Operator
         u_n = self.elastic_normal_fracture_deformation(subdomains)
         return tangential_to_nd @ u_t + normal_to_nd @ u_n
 
+    @pp.ad.cached_method
```
Likely no, I blame it on momentum after caching the two above methods. I think the best option is to roll back and also reintroduce the typing; we have said that caching should be done sparingly. Agree?

IvarStefansson replied:
Yes

keileg commented on src/porepy/models/protocol.py:
```diff
@@ -918,6 +918,89 @@ def load_data_from_pvd(
+        with @pp.ad.cached_method, and that are not decleared in the other protocols.
+        """
+
+        # This is a workaround for a technical issue: While the decorated methods should
```
My guiding principle is not to put technical notes in docstrings, but I'm not sure that is wise.

keileg commented:
@IvarStefansson I added some tests (using a Mock class rather than the actual constitutive laws, but that should be equivalent). Please have a look, and consider in particular if there are extra cases that should be tested. I believe the setup covers relevant scenarios related to mixins, but I am not 100% sure.

Also removed the decorator from the `plastic_displacement_jump` method and improved documentation.

IvarStefansson reviewed (approved):
Looks good to me.
